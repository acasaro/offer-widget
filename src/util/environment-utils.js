import { get } from 'lodash';
import { API, APP, ENDPOINTS } from './constants/environment';

/**
 * getEnvironment - Returns the current environment, or development by default
 * @returns {String}
 */
export const getEnvironment = () => {
	if (process.env.NODE_ENV) {
		return process.env.NODE_ENV;
	}

	return 'development';
};

/**
 * userLayer7 - Whether or not to use Layer7 protected endpoints
 * @returns {Boolean}
 */
export const useLayer7 = () => {
	const environment = getEnvironment();
	return environment === 'stage' || environment === 'production';
};

/**
 * getApiUrl  - Returns the URL for the api, given the current environment
 * @returns {String}
 */
export const getApiUrl = () => {
	switch (getEnvironment()) {
		case 'production':
			return API.PRODUCTION;
		case 'stage':
			return API.STAGE;
		case 'test':
			return API.TEST;
		case 'development':
		default:
			return API.DEV;
	}
};

/**
 * getAppUrl  - Returns the URL for the app, given the environment
 * @returns {String}
 */
export const getAppUrl = () => {
	switch (getEnvironment()) {
		case 'production':
			return APP.PRODUCTION;
		case 'stage':
			return APP.STAGE;
		case 'test':
			return APP.TEST;
		case 'development':
		default:
			return APP.DEV;
	}
};

/**
 * getEndpoint  - Returns the endpoint based on environment
 * @returns {String}
 */
export const getEndpoint = service => get(ENDPOINTS, `${service}.${useLayer7() ? 'layer7' : 'insecure'}`);

/**
 * getMockData  - Returns the filepath for given mock data file
 * @param {String} fileName name of file to find path for
 * @returns {String}
 */
export const getMockData = fileName => {
	const isLocal = window.location.href.includes('localhost');
	return isLocal
		? `/mock-data/${fileName}`
		: `/etc/designs/digitaladvocacy/clientlibs/homepage-ui/mock-data/${fileName}`;
};
