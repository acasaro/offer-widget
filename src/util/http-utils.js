import axios from 'axios';
import { get as lodashGet } from 'lodash';

/**
 * getQueryString - Get a query string parameter by name
 * Adapted from https://stackoverflow.com/questions/901115/how-can-i-get-query-string-values-in-javascript
 *
 * @param {String} name   Name of parameter to retrieve
 * @returns {{String|Null}}
 */
export const getQueryString = name => {
	const url = lodashGet(window, 'location.search');
	name = name.replace(/[[\]]/g, '\\$&');
	const regex = new RegExp(`[?&]${name}(=([^&#]*)|&|#|$)`);
	const results = regex.exec(url);

	if (!results) {
		return null;
	}

	if (!results[2]) {
		return '';
	}

	return decodeURIComponent(results[2].replace(/\+/g, ' '));
};
