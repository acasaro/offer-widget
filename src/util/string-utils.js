import React from 'react';
import { forOwn } from 'lodash';
import AnalyticsLink from '../components/general/analytics-link/analytics-link';
import { POST_TYPES } from '../util/constants/analytics';

export const REPLACEMENT_VARIABLE_OPTION_TYPES = {
	ON_CLICK: 'onClick',
	STYLED_SPAN: 'styledSpan',
};

/**
 * getExcerpt - Returns an excerpt from a longer string
 *
 * @param {String} content      Longer block of content
 * @param {Number} approxLength Number of characters desired, approximately
 * @returns {String}
 */
export const getExcerpt = (content, approxLength) => {
	if (content.length <= approxLength) {
		return content;
	}

	// Split content by words
	const contentArray = content.substr(0, approxLength).split(' ');

	// Set the last "word" in the array to an ellipsis
	contentArray[contentArray.length - 1] = '...';
	return contentArray.join(' ');
};

/**
 * getUniqueId - Generate a random HTML id
 *
 * @returns {String}
 */
export const getUniqueId = () =>
	`component_${Math.random()
		.toString(36)
		.substr(2, 6)}`;

/**
 * replaceVariables - Replaces {{variables}} from AEM with JS variables
 *
 * @param {String}  str       String to search and replace variables in
 * @param {Object}  variables Variables to replace in str, key = var name
 * @returns {String|JSX}
 */
export const replaceVariables = (str = '', variables = {}) => {
	if (!str || !str.length) {
		return '';
	}

	forOwn(variables, (value, key) => {
		const strVar = `{{${key}}}`;
		const index = str.indexOf(strVar);

		if (index > -1) {
			const keyLength = strVar.length;
			let firstHalf = '';

			if (index !== 0) {
				firstHalf = str.substr(0, index);
			}

			const lastHalf = str.substr(index + keyLength);

			// If the variable being replaced is something dynamic (an object),
			// we can return JSX for render
			if (typeof value === 'object' && Object.keys(value).length) {
				// Check for click handlers, and return a link with a handler
				if (value[REPLACEMENT_VARIABLE_OPTION_TYPES.ON_CLICK]) {
					str = (
						<span>
							{firstHalf && <span className="first-half">{firstHalf}</span>}
							<AnalyticsLink
								href="#"
								onClick={value[REPLACEMENT_VARIABLE_OPTION_TYPES.ON_CLICK]}
								className={value.anchorClass}
								type={POST_TYPES.CUSTOM_LINK}
								analyticsData={{ contentName: 'Retry Now link - api error message', contentCategory: 'error page' }}
							>
								{value.text}
							</AnalyticsLink>
							{lastHalf && <span className="second-half">{lastHalf}</span>}
						</span>
					);
					// If there needs to be a span (with a specific class) wrapping the variable,
					// we can return some JSX to handle that
				} else if (value[REPLACEMENT_VARIABLE_OPTION_TYPES.STYLED_SPAN]) {
					str = (
						<span>
							{firstHalf && <span className="first-half">{firstHalf}</span>}
							<span className={value[REPLACEMENT_VARIABLE_OPTION_TYPES.STYLED_SPAN]}>{value.text}</span>
							{lastHalf && <span className="second-half">{lastHalf}</span>}
						</span>
					);
				}
			} else {
				str = firstHalf + value + lastHalf;
			}
		}
	});

	return str;
};

/**
 * formatCurrency - Convert a number string to currency format
 *
 * Adapted from https://stackoverflow.com/questions/149055/how-can-i-format-numbers-as-dollars-currency-string-in-javascript
 *
 * @param {String}  str          String to format as currency
 * @param {String}  currencySign Symbol to prepend to currency string output
 * @returns {String}
 */
export const formatCurrency = (str, currencySign) =>
	`${currencySign}${Number.parseFloat(str)
		.toFixed(2)
		.replace(/(\d)(?=(\d{3})+\.)/g, '$1,')}`;

/**
 * trim - Trims a string if it exists, else returns blank string
 *
 * @param {String} str     String to trim spaces from
 * @returns {String}
 */
export const trim = str => (str && str.length ? str.replace(/^\s+|\s+$/gm, '') : '');

/**
 * isNullEmptyOrUndefined - Checks to see if the parameter is null,
 * empty, or undefined.
 *
 * @param {*} str     Item to verify
 * @returns {Boolean}
 */
export const isNullEmptyOrUndefined = str =>
	str === null || typeof str === 'undefined' || (typeof str === 'string' && trim(str) === '');

/**
 * toTitleCase - Converts strings to title case
 *
 * @param {String} str
 * @returns {String}
 */
export const toTitleCase = str =>
	str && str.length ? str.replace(/\w\S*/g, txt => txt.charAt(0).toUpperCase() + txt.substr(1).toLowerCase()) : '';
