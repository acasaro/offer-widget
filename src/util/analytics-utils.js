import { get } from 'lodash';
import { getEnvironment } from './environment-utils';
import { POST_TYPES, SECTION_NAMES, ANALYTICS_SCRIPT_SRC } from './constants/analytics';

// To do: strip out unnecessary stuff, only post offers. to the postDataObject, and on click.

export class AnalyticsQueue {
	// Wait a max of ten seconds
	static MAX_WAIT = 10000;

	// Check every second for publishPostPageData
	static CHECK_EVERY = 1000;

	constructor() {
		this.queue = [];
		this.runningTime = 0;
	}

	// Executes publishPostPageData immediately if possible,
	// else adds analytics data to queue for retry
	queueOrExecute = (type, content) => {
		if (window.publishPostPageData) {
			window.publishPostPageData(type, content);
		} else {
			this.queue.push({ type, content });
			// If the analytics queue hasn't started processing, start it
			if (this.runningTime === 0) {
				this.executeQueue();
			}
		}
	};

	executeQueue = () => {
		// If publishPostPageData is available, send all queued
		// items to Adobe analytics, then clear the queue
		if (window.publishPostPageData) {
			this.queue.forEach(item => {
				window.publishPostPageData(item.type, item.content);
			});

			this.queue = [];
		} else if (this.runningTime < AnalyticsQueue.MAX_WAIT) {
			this.runningTime += AnalyticsQueue.CHECK_EVERY;
			window.setTimeout(this.executeQueue, AnalyticsQueue.CHECK_EVERY);
		}
	};
}

// Instantiate analytics queue immediately, but don't start timers
const analyticsQueue = new AnalyticsQueue();

// TODO: make businessUnit & website parameters, rather than hardcoded

export const updatePageDataLayer = (pageName, loginStatus, userName, referringSite, referringSiteSection) => {
	const updatedPageDataLayer = {
		content: {
			pageName,
			referringSite,
			referringSiteSection,
			siteSectionL1: '',
			siteSectionL2: '',
			siteSectionL3: '',
			businessUnit: 'optum',
			website: 'myoptum',
		},
		user: {
			loginStatus,
			userType: 'consumer',
		},
		dacoData: {
			protected: {
				user: {
					userName,
					userType: 'consumer',
					login: loginStatus,
				},
			},
		},
	};

	window.pageDataLayer = { ...window.pageDataLayer, ...updatedPageDataLayer };
};

export const updateTracking = eligibility => {
	const href = window.location.href;
	const referringSiteSection = '';
	let loginStatus = '';
	let userName = '';
	let referringSite = '';
	loginStatus = 'not loggedin';
	let page = window.location.pathname.substring(window.location.pathname.lastIndexOf('/') + 1);
	let pageName = page.replace('.html', '');

	updatePageDataLayer(pageName, loginStatus, userName, referringSite, referringSiteSection);
};

/**
 * injectAnalyticsScript  - Appends environment-specific satellite library script to document head, calls _satellite.pageBottom()
 */
export const injectAnalyticsScript = async () => {
	const satelliteLibScript = document.createElement('script');
	satelliteLibScript.async = true;
	satelliteLibScript.id = 'satLibScript';
	satelliteLibScript.src = getEnvironment() === 'production' ? ANALYTICS_SCRIPT_SRC.PROD : ANALYTICS_SCRIPT_SRC.DEFAULT;

	document.head.appendChild(satelliteLibScript);

	// Page needs analytics. Wait up to 2 seconds for analytics to have loaded
	// before bailing
	const CHECK_EVERY = 50;
	const MAX_WAIT = 2000;
	let timer = 0;
	return new Promise(resolve => {
		const loadInterval = window.setInterval(() => {
			if (window._satellite && window._satellite.pageBottom) {
				window.clearInterval(loadInterval);

				// If no pageDataLayer is present, create one
				if (!window.pageDataLayer) {
					updateTracking();
				}
				// If no publishPostPageData available, initialize the analytics script
				if (!window.publishPostPageData) {
					window._satellite.pageBottom();
				}
				resolve(true);
			} else if (timer > MAX_WAIT) {
				window.clearInterval(loadInterval);
				resolve(false);
			} else {
				timer += CHECK_EVERY;
			}
		}, CHECK_EVERY);
	});
};

export const postAnalyticsData = (type, content) => {
	const postObject = {
		user: get(window, 'pageDataLayer.user'),
		content: content,
	};

	// Either execute or add the current analytics post to the queue
	analyticsQueue.queueOrExecute(type, postObject);

	return analyticsQueue.queue.length === 0;
};
