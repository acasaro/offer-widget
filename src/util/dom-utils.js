import React from 'react';
import { render, unmountComponentAtNode } from 'react-dom';
import { get } from 'lodash';
import { isNullEmptyOrUndefined } from './string-utils';

export const checkIfElementExists = selector => {
	const element = document.querySelector(selector);
	if (element !== null) {
		return true;
	}
	return false;
};

export const renderIfExists = (Component, selector, opts = {}) => {
	if (checkIfElementExists(selector)) {
		render(Component, document.querySelector(selector));
		return true;
	}
	// If mounting was unsuccessful, return false and put up a console warning
	console.warn(`Attempting to mount to ${selector} failed. DOM node not found.`);
	return false;
};

/**
 * linkTarget  - takes in boolean, returns '_blank' for true, '_self' for false
 * @param {Boolean}  openNewWindow       Whether to open in new tab or not
 * @returns {String}
 */
export const linkTarget = openNewWindow => (openNewWindow ? '_blank' : '_self');

export const getTabbables = element => {
	const tabbableQuery = ['input', 'select', 'a[href]', 'textarea', 'button', '[role=button]', '[tabindex]'];

	return element.querySelectorAll(tabbableQuery.join(','));
};
