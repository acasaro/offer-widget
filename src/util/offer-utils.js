import { get, forOwn } from 'lodash';
import { isNullEmptyOrUndefined } from './string-utils';
import { getApiUrl, getEndpoint, getMockData } from './environment-utils';
import { useMockData } from '../initialization-helper';

export const sortOffers = (offers = []) =>
	offers.sort((a, b) => {
		// Prioritize AEM sort order if applicable, otherwise use OGM-specified sort order
		const aOrder = get(a, 'content.order', a.order);
		const bOrder = get(b, 'content.order', b.order);

		return parseInt(aOrder, 10) - parseInt(bOrder, 10);
	});

/**
 * hasModal - checks if object has non-null keys
 * @param {Object} modalArray
 * @returns {Boolean}
 */
export const hasModal = modalArray => {
	// default false
	let hasModal = false;
	forOwn(modalArray, (value, key) => {
		// if any of the modal keys are not null, return true
		if (!isNullEmptyOrUndefined(value)) {
			hasModal = true;
		}
	});

	return hasModal;
};

/**
 * getOffersUrl - should return offers GET url or mock data, depending on mock data
 */

// TODO: parameterize the query strings depending on the portal making the request
export const getOffersUrl = () => {
	return useMockData()
		? getMockData('offers-mock-data.json')
		: getApiUrl() + '/' + getEndpoint('leadInteraction') + '?portal=homepage&language=en';
};
