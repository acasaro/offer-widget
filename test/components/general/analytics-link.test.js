import React from 'react';
import { mount } from 'enzyme';
import AnalyticsLink from '../../../src/components/general/analytics-link/analytics-link';
import { postAnalyticsData } from '../../../src/util/analytics-utils';

jest.mock('../../../src/util/analytics-utils', () => ({
	postAnalyticsData: jest.fn(),
}));

describe('Analytics link component', () => {
	beforeEach(() => {
		process.env.NODE_ENV = 'test';
	});

	it('should correctly render a component', () => {
		const wrapper = mount(<AnalyticsLink />);
		expect(wrapper).toMatchSnapshot();
	});

	it('should call the original click handler when an analytics link is clicked', () => {
		const mockOnClick = jest.fn();
		const wrapper = mount(<AnalyticsLink onClick={mockOnClick} />);
		postAnalyticsData.mockClear();
		wrapper.simulate('click');
		expect(mockOnClick).toHaveBeenCalled();
		expect(postAnalyticsData).toHaveBeenCalled();
	});

	it('should call postAnalyticsData but not an onClick if no onClick is passed in', () => {
		const wrapper = mount(<AnalyticsLink />);
		postAnalyticsData.mockClear();
		wrapper.simulate('click');
		expect(postAnalyticsData).toHaveBeenCalled();
	});

	it('should not call event.preventDefault if there is an href provided', () => {
		const wrapper = mount(<AnalyticsLink href="http://something.com" />);
		const event = {
			preventDefault: jest.fn(),
		};
		wrapper.instance().handleClick(event);
		expect(event.preventDefault).not.toHaveBeenCalled();
	});

	afterEach(() => {
		// Set environment back to test
		process.env.NODE_ENV = 'test';
	});
});
