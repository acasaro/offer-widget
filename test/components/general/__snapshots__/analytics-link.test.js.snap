// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analytics link component should correctly render a component 1`] = `
<AnalyticsLink>
  <a
    alt={undefined}
    aria-controls={undefined}
    aria-expanded={undefined}
    aria-label={undefined}
    className={undefined}
    href="#"
    id={undefined}
    onBlur={undefined}
    onClick={[Function]}
    onFocus={undefined}
    onMouseEnter={undefined}
    onMouseLeave={undefined}
    role={undefined}
    style={undefined}
    target={undefined}
  />
</AnalyticsLink>
`;
